pkg/agent/nodeportlocal/portcache/port_table.go

```diff
@@ -85,25 +89,47 @@ func (pt *PortTable) GetDataForPodIP(ip string) []NodePortData {
 	return allData
 }
 
-func (pt *PortTable) GetEntryByPodIPPort(ip string, port int) *NodePortData {
+func (pt *PortTable) GetEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
-	return pt.getEntryByPodIPPort(ip, port)
+	return pt.getEntryByPodIPPortProtocol(ip, port, protocol)
 }
 
-func (pt *PortTable) getEntryByPodIPPort(ip string, port int) *NodePortData {
+func (pt *PortTable) getEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
 	for _, data := range pt.Table {
-		if data.PodIP == ip && data.PodPort == port {
+		if data.PodIP == ip && data.PodPort == port && data.Protocol == protocol {
 			return &data
 		}
 	}
 	return nil
 }
 
-func (pt *PortTable) getFreePort(podIP string, podPort int) (int, Closeable, error) {
+func (pt *PortTable) isNodePortClaimedForPodIPProtocol(ip string, nodeport int, protocol string) bool {
+	var val NodePortData
+	ok := false
+	// Since NPL supports only TCP and UDP protocols, we can check if an entry corresponding to key `<nodeport>:udp`
+	// is present in the PortTable, if the protocol we want to serve is TCP, and vice versa.
+	switch protocol {
+	case strings.ToLower(string(corev1.ProtocolTCP)):
+		val, ok = pt.Table[util.BuildPortProto(fmt.Sprint(nodeport), strings.ToLower(string(corev1.ProtocolUDP)))]
```

Comment:
I believe `BuildPortProto` already calls `strings.ToLower` on the protocol parameter